File: Chapter2/Codes_Empirical_Examples/all_examples_VAR.R
# =============================================================================
# VAR EMPIRICAL EXAMPLES FOR MACROECONOMETRICS TEXTBOOK
# Chapter: VAR Models
# =============================================================================
#
# Textbook: Macroeconometrics
# Author: Alessia Paccagnini
#
#
# Data Source: FRED St. Louis McCracken & Ng Dataset (2026-01-QD.xlsx)
#
# This script provides three comprehensive examples using real quarterly US data:
# 1. VAR Estimation with Information Criteria, Residual Diagnostics, and Granger Causality
# 2. VECM Example with Cointegration Testing (Consumption-Income)
# 3. Reduced-Form Impulse Response Functions (motivating the need for identification)
#
# Period: 1960:Q2 to 2026:Q1 (264 observations for VAR, 265 for cointegration)
# Updated: March 3, 2026
#
# Required packages: readxl, vars, urca, tseries, lmtest
# Install with: install.packages(c("readxl", "vars", "urca", "tseries", "lmtest"))
# =============================================================================

library(readxl)
library(vars)
library(urca)
library(tseries)
library(lmtest)

cat(strrep("=", 80), "\n")
cat("MACROECONOMETRICS TEXTBOOK: VAR EMPIRICAL EXAMPLES\n")
cat("Author: Alessia Paccagnini, University College Dublin\n")
cat("Using REAL DATA: FRED McCracken-Ng Dataset (2026-01-QD.xlsx)\n")
cat(strrep("=", 80), "\n\n")

# ============================================================================
# FILE PATH — adjust for your environment
# ============================================================================

# --- Option A: Interactive file picker (uncomment this line) ---
# file_path <- file.choose()  # Opens a dialog to select the Excel file

# --- Option B: Set path manually ---
file_path <- "2026-01-QD.xlsx"

# ============================================================================
# LOAD DATA
# ============================================================================
# File structure (sheet 'in'):
#   Row 1: column names (sasdate, GDPC1, CPIAUCSL, ...)
#   Row 2: factor labels
#   Row 3: transformation codes
#   Rows 4-5: 1959:Q1-Q2 data
#   Row 6+: 1959:Q3 onward — we start here
#
# read_excel with col_names = FALSE reads row 1 as data.
# Row 6 in the file = row 6 in R (1-indexed) when col_names = FALSE.
# This matches Python's iloc[5] (0-indexed).
# ============================================================================

raw <- read_excel(file_path, sheet = "in", col_names = FALSE)
colnames_vec <- as.character(raw[1, ])

gdp_col    <- which(colnames_vec == "GDPC1")
cpi_col    <- which(colnames_vec == "CPIAUCSL")
ff_col     <- which(colnames_vec == "FEDFUNDS")
cons_col   <- which(colnames_vec == "PCECC96")
income_col <- which(colnames_vec == "DPIC96")

# Start from row 6 (= 1959:Q3, matching Python iloc[5])
start_row <- 6

gdp    <- as.numeric(raw[[gdp_col]][start_row:nrow(raw)])
cpi    <- as.numeric(raw[[cpi_col]][start_row:nrow(raw)])
ff     <- as.numeric(raw[[ff_col]][start_row:nrow(raw)])
cons   <- as.numeric(raw[[cons_col]][start_row:nrow(raw)])
income <- as.numeric(raw[[income_col]][start_row:nrow(raw)])

n_raw <- length(gdp)

# Annualised quarter-on-quarter growth rates
gdp_growth <- c(NA, 100 * 4 * diff(log(gdp)))
inflation  <- c(NA, 100 * 4 * diff(log(cpi)))

# Remove first observation (NaN from differencing) and build macro data
macro_mat <- data.frame(
  GDP_Growth = gdp_growth[-1],
  Inflation  = inflation[-1],
  FedFunds   = ff[-1]
)

# Remove any rows with NA
macro_mat <- macro_mat[complete.cases(macro_mat), ]

# Cointegration data — full series, no differencing needed
coint_mat <- data.frame(
  LogConsumption = log(cons),
  LogIncome      = log(income)
)
coint_mat <- coint_mat[complete.cases(coint_mat), ]

cat(sprintf("Macro data: %d observations (1960:Q2 to 2026:Q1)\n", nrow(macro_mat)))
cat(sprintf("Cointegration data: %d observations\n\n", nrow(coint_mat)))

# Create ts objects
var_ts <- ts(macro_mat, start = c(1960, 2), frequency = 4)

# ============================================================================
# EXAMPLE 1: VAR ESTIMATION
# ============================================================================

cat(strrep("=", 80), "\n")
cat("EXAMPLE 1: VAR ESTIMATION, DIAGNOSTICS, AND GRANGER CAUSALITY\n")
cat(strrep("=", 80), "\n\n")

# ---- Descriptive statistics ----
cat("Descriptive Statistics:\n")
print(summary(var_ts))

# ---- Unit root tests ----
cat("\n", strrep("-", 70), "\n")
cat("UNIT ROOT TESTS (Augmented Dickey-Fuller)\n")
cat(strrep("-", 70), "\n\n")

for (v in colnames(macro_mat)) {
  adf <- adf.test(macro_mat[[v]], alternative = "stationary")
  conclusion <- ifelse(adf$p.value < 0.05, "Stationary", "Non-stationary")
  cat(sprintf("  %-15s ADF = %7.3f  p = %5.3f  %s\n",
              v, adf$statistic, adf$p.value, conclusion))
}

# ---- Lag selection ----
cat("\n", strrep("-", 70), "\n")
cat("LAG ORDER SELECTION\n")
cat(strrep("-", 70), "\n\n")

lag_select <- VARselect(var_ts, lag.max = 12, type = "const")
print(lag_select$selection)
cat("\n")
print(round(t(lag_select$criteria[1:3, 1:8]), 3))

# NOTE: BIC(1) and BIC(2) differ by < 0.001. To ensure consistency
# across R, Python, and MATLAB, we fix p = 1 (the textbook specification).
optimal_lag <- 1
cat(sprintf("\nUsing lag order p = %d (textbook specification; BIC near-tied between p=1 and p=2)\n",
            optimal_lag))

# ---- VAR estimation ----
cat("\n", strrep("-", 70), "\n")
cat(sprintf("VAR(%d) ESTIMATION RESULTS\n", optimal_lag))
cat(strrep("-", 70), "\n\n")

var_model <- VAR(var_ts, p = optimal_lag, type = "const")
print(summary(var_model))

# ---- Residual diagnostics ----
cat("\n", strrep("-", 70), "\n")
cat("RESIDUAL DIAGNOSTICS\n")
cat(strrep("-", 70), "\n\n")

resid_mat <- residuals(var_model)

# Durbin-Watson
cat("Durbin-Watson:\n")
for (v in colnames(resid_mat)) {
  dw_val <- sum(diff(resid_mat[, v])^2) / sum(resid_mat[, v]^2)
  cat(sprintf("  %s: DW = %.3f\n", v, dw_val))
}

# Ljung-Box Q(12)
cat("\nLjung-Box Q(12):\n")
for (v in colnames(resid_mat)) {
  lb <- Box.test(resid_mat[, v], lag = 12, type = "Ljung-Box")
  sig <- ifelse(lb$p.value < 0.05, "**", "")
  cat(sprintf("  %s: Q = %.2f, p = %.4f %s\n", v, lb$statistic, lb$p.value, sig))
}

# Jarque-Bera
cat("\nJarque-Bera:\n")
for (v in colnames(resid_mat)) {
  jb <- jarque.bera.test(resid_mat[, v])
  sig <- ifelse(jb$p.value < 0.05, "**", "")
  cat(sprintf("  %s: JB = %.2f, p = %.4f %s\n", v, jb$statistic, jb$p.value, sig))
}

# Residual correlation
cat("\nResidual Correlation Matrix:\n")
corr_resid <- cor(resid_mat)
print(round(corr_resid, 3))

# ---- Granger causality ----
# Using lmtest::grangertest for pairwise bivariate tests
# Note: for exact match with Python/MATLAB (which also run bivariate tests),
# we use grangertest(response ~ predictor), which tests H0: predictor does NOT
# Granger-cause response.
cat("\n", strrep("-", 70), "\n")
cat("GRANGER CAUSALITY TESTS\n")
cat(strrep("-", 70), "\n\n")

var_names <- colnames(macro_mat)
cat(sprintf("%-45s %10s %10s\n", "Null Hypothesis", "F-stat", "p-value"))
cat(strrep("-", 70), "\n")

# Pairwise bivariate Granger tests (matches Python/MATLAB convention)
pairs <- list(
  c(1, 2), c(1, 3),  # what causes GDP_Growth? (Inflation, FedFunds)
  c(2, 1), c(2, 3),  # what causes Inflation? (GDP_Growth, FedFunds)
  c(3, 1), c(3, 2)   # what causes FedFunds? (GDP_Growth, Inflation)
)
for (p in pairs) {
  resp <- p[1]; pred <- p[2]
  # grangertest(response ~ predictor): tests H0: predictor does NOT cause response
  gc <- grangertest(var_ts[, resp] ~ var_ts[, pred], order = optimal_lag)
  f_stat <- gc$F[2]
  p_val  <- gc$`Pr(>F)`[2]
  sig <- ifelse(p_val < 0.01, "***", ifelse(p_val < 0.05, "**", ""))
  label <- sprintf("%s does not cause %s", var_names[pred], var_names[resp])
  cat(sprintf("%-45s %10.2f %10.4f %s\n", label, f_stat, p_val, sig))
}

# ---- Plot 1: Residual diagnostics ----
png("ex1_var_diagnostics.png", width = 1400, height = 1200, res = 150)
par(mfrow = c(3, 3), mar = c(4, 4, 3, 1))
for (i in 1:3) {
  v <- var_names[i]
  r <- resid_mat[, i]

  # Time series
  plot(r, type = "l", col = "blue", lwd = 0.8,
       main = paste0(v, ": Time Series"), ylab = "Residuals", xlab = "")
  abline(h = 0, col = "red", lty = 2)

  # Histogram
  hist(r, breaks = 30, probability = TRUE, col = rgb(0.27, 0.51, 0.71, 0.7),
       border = "white", main = paste0(v, ": Histogram"), xlab = "")
  curve(dnorm(x, mean(r), sd(r)), add = TRUE, col = "red", lwd = 2)

  # ACF
  acf(r, lag.max = 20, main = paste0(v, ": ACF"), col = "steelblue")
}
dev.off()
cat("\nSaved: ex1_var_diagnostics.png\n")

# ---- Plot 2: Lag selection ----
png("ex1_lag_selection.png", width = 1000, height = 600, res = 150)
ic_mat <- t(lag_select$criteria[1:3, ])
lags_vec <- 1:ncol(lag_select$criteria)
plot(lags_vec, ic_mat[, 1], type = "b", col = "green3", pch = 16, lwd = 2,
     ylim = range(ic_mat), xlab = "Number of Lags",
     ylab = "Information Criterion Value",
     main = "VAR Lag Order Selection\n(Real Data: 1960:Q2 - 2026:Q1)")
lines(lags_vec, ic_mat[, 2], type = "b", col = "red", pch = 15, lwd = 2)
lines(lags_vec, ic_mat[, 3], type = "b", col = "blue", pch = 17, lwd = 2)
abline(v = lag_select$selection["AIC(n)"], col = "green3", lty = 2)
abline(v = lag_select$selection["SC(n)"],  col = "red",    lty = 2)
abline(v = lag_select$selection["HQ(n)"],  col = "blue",   lty = 2)
legend("topright", legend = c("AIC", "BIC", "HQIC"),
       col = c("green3", "red", "blue"), pch = c(16, 15, 17), lwd = 2)
grid()
dev.off()
cat("Saved: ex1_lag_selection.png\n")

# ============================================================================
# EXAMPLE 2: COINTEGRATION
# ============================================================================

cat("\n", strrep("=", 80), "\n")
cat("EXAMPLE 2: COINTEGRATION AND VECM\n")
cat(strrep("=", 80), "\n\n")

cat(sprintf("Observations: %d\n\n", nrow(coint_mat)))

# ---- Unit root tests ----
cat("Unit Root Tests:\n")
coint_names <- c("LogConsumption", "LogIncome")
for (v in coint_names) {
  adf_lev  <- adf.test(coint_mat[[v]], alternative = "stationary")
  adf_diff <- adf.test(diff(coint_mat[[v]]), alternative = "stationary")
  cat(sprintf("  %s:\n", v))
  cat(sprintf("    Levels: ADF = %7.2f, p = %5.3f\n", adf_lev$statistic, adf_lev$p.value))
  cat(sprintf("    Diffs:  ADF = %7.2f, p = %5.3f\n", adf_diff$statistic, adf_diff$p.value))
}

# ---- Johansen cointegration test ----
# Case 4: linear trend in data, constant restricted to cointegrating equation.
# This is the correct specification when variables have deterministic trends
# (productivity growth). Using ecdet="const" (Case 2) spuriously finds r=2
# because the test confuses trend-stationarity with I(0).
cat("\n", strrep("-", 70), "\n")
cat("JOHANSEN COINTEGRATION TEST (Case 4: linear trend in data)\n")
cat(strrep("-", 70), "\n\n")

coint_ts <- ts(coint_mat, start = c(1960, 1), frequency = 4)

# urca::ca.jo with ecdet="trend" corresponds to Case 4 (det_order=1 in Python)
johansen <- ca.jo(coint_ts, type = "trace", ecdet = "trend", K = 2)
cat("Trace Test:\n")
print(summary(johansen))

johansen_eigen <- ca.jo(coint_ts, type = "eigen", ecdet = "trend", K = 2)
cat("\nMax Eigenvalue Test:\n")
print(summary(johansen_eigen))

# ---- OLS long-run relationship ----
ols_fit <- lm(LogConsumption ~ LogIncome, data = coint_mat)
ols_int   <- coef(ols_fit)[1]
ols_slope <- coef(ols_fit)[2]
cat(sprintf("\nOLS Long-run: LogC = %.4f + %.4f * LogY\n", ols_int, ols_slope))

# ---- Plot 3: Cointegration ----
dates_coint <- seq(as.Date("1960-01-01"), by = "quarter", length.out = nrow(coint_mat))

png("ex2_vecm_cointegration.png", width = 1200, height = 1000, res = 150)
par(mfrow = c(3, 1), mar = c(4, 4, 3, 1))

# Time series
plot(dates_coint, coint_mat$LogConsumption, type = "l", col = "blue", lwd = 1.5,
     main = "Time Series of Log Consumption and Log Income",
     ylab = "Log Scale", xlab = "")
lines(dates_coint, coint_mat$LogIncome, col = "orange", lwd = 1.5)
legend("topleft", legend = c("Log Consumption", "Log Income"),
       col = c("blue", "orange"), lwd = 1.5)
grid()

# Scatter plot with OLS line
plot(coint_mat$LogIncome, coint_mat$LogConsumption,
     pch = 20, col = rgb(0.27, 0.51, 0.71, 0.4), cex = 0.5,
     main = "Long-run Relationship",
     xlab = "Log Income", ylab = "Log Consumption")
abline(ols_fit, col = "red", lwd = 2)
legend("topleft",
       legend = sprintf("Long-run: C = %.2f + %.3f × Y", ols_int, ols_slope),
       col = "red", lwd = 2)
grid()

# Cointegrating residual
coint_resid <- residuals(ols_fit)
coint_resid_norm <- (coint_resid - mean(coint_resid)) / sd(coint_resid)
plot(dates_coint, coint_resid_norm, type = "l", col = "blue", lwd = 1,
     main = "Cointegrating Residual (Standardized)",
     xlab = "Quarter", ylab = "Deviation from Equilibrium")
abline(h = 0, col = "red", lty = 2)
polygon(c(dates_coint, rev(dates_coint)),
        c(rep(0, length(dates_coint)), rev(coint_resid_norm)),
        col = rgb(0, 0, 1, 0.2), border = NA)
grid()

dev.off()
cat("Saved: ex2_vecm_cointegration.png\n")

# ============================================================================
# EXAMPLE 3: IDENTIFICATION PROBLEM
# ============================================================================

cat("\n", strrep("=", 80), "\n")
cat("EXAMPLE 3: REDUCED-FORM IRFs AND IDENTIFICATION PROBLEM\n")
cat(strrep("=", 80), "\n\n")

cat("Residual Correlation Matrix:\n")
print(round(corr_resid, 3))

# ---- Plot 4: Correlation heatmap ----
png("ex3_innovation_correlation.png", width = 800, height = 600, res = 150)
par(mar = c(5, 5, 4, 6))

# Simple heatmap using image()
cols <- colorRampPalette(c("blue", "white", "red"))(100)
image(1:3, 1:3, corr_resid[3:1, ], col = cols, zlim = c(-1, 1),
      axes = FALSE, xlab = "", ylab = "",
      main = "Correlation of Reduced-Form Innovations\n(The Root of the Identification Problem)")
axis(1, at = 1:3, labels = var_names, las = 1)
axis(2, at = 1:3, labels = rev(var_names), las = 2)
box()

# Add text labels
for (i in 1:3) {
  for (j in 1:3) {
    val <- corr_resid[4 - j, i]
    clr <- ifelse(abs(val) > 0.5, "white", "black")
    text(i, j, sprintf("%.3f", val), col = clr, font = 2, cex = 1.2)
  }
}

dev.off()
cat("Saved: ex3_innovation_correlation.png\n")

# ---- Plot 5: Reduced-form IRFs ----
irf_result <- irf(var_model, n.ahead = 20, ortho = FALSE, boot = FALSE)

png("ex3_reduced_form_irf.png", width = 1400, height = 1200, res = 150)
par(mfrow = c(3, 3), mar = c(4, 4, 3, 1))

for (i in seq_along(var_names)) {   # impulse
  for (j in seq_along(var_names)) { # response
    imp_name <- var_names[i]
    resp_name <- var_names[j]

    vals <- irf_result$irf[[imp_name]][, resp_name]
    horizons <- 0:(length(vals) - 1)

    plot(horizons, vals, type = "l", col = "blue", lwd = 2,
         main = paste0("Response of ", resp_name, "\nto ", imp_name, " innovation"),
         xlab = ifelse(j == 3, "Quarters", ""), ylab = "Response")
    abline(h = 0, col = "black", lwd = 0.8)
    polygon(c(horizons, rev(horizons)),
            c(rep(0, length(horizons)), rev(vals)),
            col = rgb(0, 0, 1, 0.2), border = NA)
    grid()
  }
}

dev.off()
cat("Saved: ex3_reduced_form_irf.png\n")

# ============================================================================
# SUMMARY
# ============================================================================
cat("\n", strrep("=", 80), "\n")
cat("ANALYSIS COMPLETE!\n")
cat(strrep("=", 80), "\n\n")
cat("Author: Alessia Paccagnini\n")
cat("University College Dublin, Smurfit Business School\n\n")
cat("Generated files:\n")
cat("  - ex1_var_diagnostics.png\n")
cat("  - ex1_lag_selection.png\n")
cat("  - ex2_vecm_cointegration.png\n")
cat("  - ex3_innovation_correlation.png\n")
cat("  - ex3_reduced_form_irf.png\n")

File: Chapter2/Codes_Empirical_Examples/all_examples_VAR.py
"""
=============================================================================
VAR EMPIRICAL EXAMPLES FOR MACROECONOMETRICS TEXTBOOK
Chapter: VAR Models
=============================================================================

Textbook: Macroeconometrics
Author: Alessia Paccagnini


Data Source: FRED St. Louis McCracken & Ng Dataset (2026-01-QD.xlsx)

This script provides three comprehensive examples using real quarterly US data:
1. VAR Estimation with Information Criteria, Residual Diagnostics, and Granger Causality
2. VECM Example with Cointegration Testing (Consumption-Income)
3. Reduced-Form Impulse Response Functions (motivating the need for identification)

Period: 1960:Q2 to 2026:Q1 (264 observations for VAR, 265 for cointegration)
Updated: March 3, 2026

IMPORTANT — Data loading:
  The Excel file has: row 0 = column names, rows 1-2 = metadata, rows 3-4 = 1959 Q1-Q2.
  We start from iloc[5] (= 1959:Q3). After computing quarterly growth rates and dropping
  the first NaN, the VAR sample begins at 1960:Q2.
=============================================================================
"""

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')  # Remove this line if running interactively
import matplotlib.pyplot as plt
from statsmodels.tsa.api import VAR, VECM
from statsmodels.tsa.vector_ar.vecm import coint_johansen
from statsmodels.tsa.stattools import adfuller, grangercausalitytests
from statsmodels.stats.diagnostic import acorr_ljungbox
from statsmodels.stats.stattools import durbin_watson
from statsmodels.graphics.tsaplots import plot_acf
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

plt.style.use('seaborn-v0_8-whitegrid')
plt.rcParams.update({'figure.figsize': (12, 8), 'font.size': 10, 'figure.dpi': 150})

# ============================================================================
# FILE PATH — adjust for your environment
# ============================================================================

# --- Option A: Google Colab (uncomment these 3 lines) ---
# from google.colab import files
# uploaded = files.upload()  # This will prompt you to select the file
# file_path = list(uploaded.keys())[0]

# --- Option B: Local / other environment ---
file_path = "2026-01-QD.xlsx"

# ============================================================================
# LOAD DATA
# ============================================================================
df = pd.read_excel(file_path, sheet_name='in', header=None)
colnames = df.iloc[0].values

gdp_idx    = np.where(colnames == 'GDPC1')[0][0]
cpi_idx    = np.where(colnames == 'CPIAUCSL')[0][0]
ff_idx     = np.where(colnames == 'FEDFUNDS')[0][0]
cons_idx   = np.where(colnames == 'PCECC96')[0][0]
income_idx = np.where(colnames == 'DPIC96')[0][0]

# iloc[5:] starts at 1959:Q3 in the raw file
gdp    = pd.to_numeric(df.iloc[5:, gdp_idx].values,    errors='coerce')
cpi    = pd.to_numeric(df.iloc[5:, cpi_idx].values,    errors='coerce')
ff     = pd.to_numeric(df.iloc[5:, ff_idx].values,     errors='coerce')
cons   = pd.to_numeric(df.iloc[5:, cons_idx].values,   errors='coerce')
income = pd.to_numeric(df.iloc[5:, income_idx].values, errors='coerce')

dates = pd.date_range(start='1960-01-01', periods=len(gdp), freq='QE')

# Annualised quarter-on-quarter growth rates
temp_df    = pd.DataFrame({'GDP': gdp, 'CPI': cpi, 'FF': ff}, index=dates)
gdp_growth = 100 * 4 * (np.log(temp_df['GDP']) - np.log(temp_df['GDP'].shift(1)))
inflation  = 100 * 4 * (np.log(temp_df['CPI']) - np.log(temp_df['CPI'].shift(1)))

macro_data = pd.DataFrame(
    {'GDP_Growth': gdp_growth, 'Inflation': inflation, 'FedFunds': ff},
    index=dates).dropna()

# Cointegration data — full series, no differencing needed
coint_data = pd.DataFrame(
    {'LogConsumption': np.log(cons), 'LogIncome': np.log(income)},
    index=dates).dropna()

var_data = macro_data[['GDP_Growth', 'Inflation', 'FedFunds']].dropna()

first_q = f"{var_data.index[0].year}:Q{var_data.index[0].quarter}"
last_q  = f"{var_data.index[-1].year}:Q{var_data.index[-1].quarter}"

print("="*80)
print(f"Sample: {first_q} to {last_q}  |  VAR obs: {len(var_data)}  |  Coint obs: {len(coint_data)}")
print("="*80)

# ============================================================================
# EXAMPLE 1 — VAR
# ============================================================================
model      = VAR(var_data)
lag_results = model.select_order(maxlags=12)

# NOTE: BIC(1)=3.835 and BIC(2)=3.834 differ by only 0.0006.
# Due to floating-point differences between MATLAB and Python in det()/log(),
# MATLAB may select p=2 while Python selects p=1. To ensure consistency
# across platforms, we fix p=1 (the textbook specification).
optimal_lag = 1  # Textbook specification; BIC near-tied between p=1 and p=2
var_result  = model.fit(optimal_lag)
residuals   = var_result.resid

print(var_result.summary())

# Granger causality
# grangercausalitytests(data[[X, Y]]) tests: does Y Granger-cause X?
print("\nGranger Causality (lag = 1):")
for x_var in var_data.columns:
    for y_var in var_data.columns:
        if x_var == y_var: continue
        gc = grangercausalitytests(var_data[[x_var, y_var]],
                                   maxlag=optimal_lag, verbose=False)
        f_stat, p_val, _, _ = gc[optimal_lag][0]['ssr_ftest']
        sig = "***" if p_val < 0.01 else ("**" if p_val < 0.05 else "")
        print(f"  {y_var:>12s} -> {x_var:<12s}  F = {float(f_stat):7.2f}  p = {float(p_val):.4f} {sig}")

# ---- Plot 1: Residual diagnostics ----
fig, axes = plt.subplots(3, 3, figsize=(14, 12))
fig.suptitle('VAR Residual Diagnostics\n(GDP Growth, Inflation, Federal Funds Rate)',
             fontsize=13, fontweight='bold')
for i, col in enumerate(var_data.columns):
    r = residuals.iloc[:, i]
    axes[i,0].plot(r.index, r.values, 'b-', lw=.8)
    axes[i,0].axhline(0, color='red', lw=1, ls='--')
    axes[i,0].set_title(f'{col}: Time Series', fontsize=10); axes[i,0].set_ylabel('Residuals')
    axes[i,1].hist(r.values, bins=30, density=True, alpha=.7, color='steelblue', edgecolor='white')
    mu, sig = r.mean(), r.std()
    xg = np.linspace(mu-4*sig, mu+4*sig, 100)
    axes[i,1].plot(xg, stats.norm.pdf(xg, mu, sig), 'r-', lw=2, label='Normal')
    axes[i,1].set_title(f'{col}: Histogram', fontsize=10); axes[i,1].legend()
    plot_acf(r, lags=20, ax=axes[i,2], color='steelblue',
             vlines_kwargs={'colors':'steelblue'})
    axes[i,2].set_title(f'{col}: ACF', fontsize=10)
plt.tight_layout()
plt.savefig('ex1_var_diagnostics.png', dpi=150, bbox_inches='tight'); plt.close()

# ---- Plot 2: Lag selection ----
fig, ax = plt.subplots(figsize=(10, 6))
lags_r = range(1, 13)
aic_v = [model.fit(p).aic  for p in lags_r]
bic_v = [model.fit(p).bic  for p in lags_r]
hq_v  = [model.fit(p).hqic for p in lags_r]
ax.plot(lags_r, aic_v, 'g-o', lw=2, ms=6, label='AIC')
ax.plot(lags_r, bic_v, 'r-s', lw=2, ms=6, label='BIC')
ax.plot(lags_r, hq_v,  'b-^', lw=2, ms=6, label='HQIC')
for x, c in [(lag_results.aic,'green'),(lag_results.bic,'red'),(lag_results.hqic,'blue')]:
    ax.axvline(x, color=c, lw=1.5, ls='--', alpha=.7)
ax.set_xlabel('Number of Lags'); ax.set_ylabel('Information Criterion Value')
ax.set_title(f'VAR Lag Order Selection\n(Real Data: {first_q} – {last_q})',
             fontsize=12, fontweight='bold')
ax.legend(); ax.grid(True, alpha=.3)
plt.tight_layout()
plt.savefig('ex1_lag_selection.png', dpi=150, bbox_inches='tight'); plt.close()

# ============================================================================
# EXAMPLE 2 — COINTEGRATION
# ============================================================================
vecm_data = coint_data[['LogConsumption','LogIncome']].dropna()

# Johansen test with det_order=1 (Case 4: linear trend in data, constant in CE)
# This is the correct specification when variables have deterministic trends
# (productivity growth). Using det_order=0 spuriously finds r=2 because the
# test confuses trend-stationarity with I(0).
johansen = coint_johansen(vecm_data, det_order=1, k_ar_diff=1)

print("\nJohansen Trace Test (Case 4: linear trend in data):")
for i in range(2):
    print(f"  r<={i}: stat={johansen.lr1[i]:.2f}  "
          f"90%={johansen.cvt[i,0]:.2f}  95%={johansen.cvt[i,1]:.2f}  99%={johansen.cvt[i,2]:.2f}")
print("\nJohansen Max Eigenvalue Test:")
for i in range(2):
    print(f"  r={i}: stat={johansen.lr2[i]:.2f}  "
          f"90%={johansen.cvm[i,0]:.2f}  95%={johansen.cvm[i,1]:.2f}  99%={johansen.cvm[i,2]:.2f}")

# OLS long-run relationship (avoids VECM beta sign ambiguity)
X_ols = np.column_stack([np.ones(len(vecm_data)), vecm_data['LogIncome'].values])
b_ols = np.linalg.lstsq(X_ols, vecm_data['LogConsumption'].values, rcond=None)[0]
ols_int, ols_slope = b_ols
print(f"OLS long-run: LogC = {ols_int:.4f} + {ols_slope:.4f} * LogY")

# ---- Plot 3: Cointegration ----
fig, axes = plt.subplots(3, 1, figsize=(12, 10))
fig.suptitle(f'Cointegration Analysis: Consumption and Income\n(Real Data: {first_q} – {last_q})',
             fontsize=13, fontweight='bold')
axes[0].plot(vecm_data.index, vecm_data['LogConsumption'], lw=1.5, label='Log Consumption')
axes[0].plot(vecm_data.index, vecm_data['LogIncome'], lw=1.5, color='orange', label='Log Income')
axes[0].set_title('Time Series of Log Consumption and Log Income'); axes[0].set_ylabel('Log Scale')
axes[0].legend(); axes[0].grid(True, alpha=.3)

xd, yd = vecm_data['LogIncome'].values, vecm_data['LogConsumption'].values
axes[1].scatter(xd, yd, alpha=.4, s=10, color='steelblue')
axes[1].plot(xd, ols_int + ols_slope*xd, 'r-', lw=2,
             label=f'Long-run: C = {ols_int:.2f} + {ols_slope:.3f} × Y')
axes[1].set_xlabel('Log Income'); axes[1].set_ylabel('Log Consumption')
axes[1].set_title('Long-run Relationship'); axes[1].legend(); axes[1].grid(True, alpha=.3)

cr = yd - (ols_int + ols_slope*xd)
cr_n = (cr - cr.mean()) / cr.std()
axes[2].plot(vecm_data.index, cr_n, 'b-', lw=1)
axes[2].axhline(0, color='red', lw=1, ls='--')
axes[2].fill_between(vecm_data.index, 0, cr_n, alpha=.2)
axes[2].set_title('Cointegrating Residual (Standardized)')
axes[2].set_xlabel('Quarter'); axes[2].set_ylabel('Deviation from Equilibrium')
axes[2].grid(True, alpha=.3)
plt.tight_layout()
plt.savefig('ex2_vecm_cointegration.png', dpi=150, bbox_inches='tight'); plt.close()

# ============================================================================
# EXAMPLE 3 — IDENTIFICATION PROBLEM
# ============================================================================
irf_obj = var_result.irf(periods=20)
corr_resid = residuals.corr()
vn = ['GDP_Growth','Inflation','FedFunds']

# ---- Plot 4: Correlation heatmap ----
fig, ax = plt.subplots(figsize=(8, 6))
im = ax.imshow(corr_resid.values, cmap='RdBu_r', vmin=-1, vmax=1, aspect='auto')
ax.set_xticks(range(3)); ax.set_yticks(range(3))
ax.set_xticklabels(vn, fontsize=11); ax.set_yticklabels(vn, fontsize=11)
for i in range(3):
    for j in range(3):
        clr = 'white' if abs(corr_resid.iloc[i,j]) > .5 else 'black'
        ax.text(j, i, f'{corr_resid.iloc[i,j]:.3f}', ha='center', va='center',
                color=clr, fontsize=14, fontweight='bold')
ax.set_title('Correlation of Reduced-Form Innovations\n(The Root of the Identification Problem)',
             fontsize=12, fontweight='bold')
plt.colorbar(im, ax=ax, label='Correlation Coefficient')
plt.tight_layout()
plt.savefig('ex3_innovation_correlation.png', dpi=150, bbox_inches='tight'); plt.close()

# ---- Plot 5: Reduced-form IRFs ----
fig, axes = plt.subplots(3, 3, figsize=(14, 12))
fig.suptitle('Reduced-Form Impulse Response Functions\n'
             '\u26A0\uFE0F NOT Economically Identified \u2014 These Are Statistical Responses Only',
             fontsize=13, fontweight='bold')
hor = np.arange(21)
for i, imp in enumerate(vn):
    for j, resp in enumerate(vn):
        ax = axes[j, i]; v = irf_obj.irfs[:, j, i]
        ax.plot(hor, v, 'b-', lw=2); ax.axhline(0, color='black', lw=.8)
        ax.fill_between(hor, 0, v, alpha=.2, color='blue')
        ax.set_title(f'Response of {resp}\nto {imp} innovation', fontsize=9)
        if j == 2: ax.set_xlabel('Quarters')
        ax.set_ylabel('Response'); ax.grid(True, alpha=.3)
fig.text(.5, .01,
         'WARNING: These are REDUCED-FORM IRFs to correlated innovations. '
         'They conflate multiple structural effects and cannot be given causal interpretation.',
         ha='center', fontsize=10, style='italic', color='darkred',
         bbox=dict(boxstyle='round', facecolor='lightyellow', alpha=.9))
plt.tight_layout(rect=[0, .04, 1, .96])
plt.savefig('ex3_reduced_form_irf.png', dpi=150, bbox_inches='tight'); plt.close()

print("\nAll 5 plots saved. Done.")
